### DISCUSSION TITLE
Mutable Data Borrow in Rust Function

### DISCUSSION THREAD

**Discussion by DasLixou** [mlua-rs/mlua#288] [Q&A] [answered]
Hello! I made a rust function and made it accessible to my lua script by setting it as a function in globals.
So what I want to do is register callback with a name, store them and then later call the functions I want from the given name.
This currently looks like this:
```lua
feature("my_feature", function ()

end)
```
```rust
fn feature(_: &Lua, (name, callback): (String, LuaFunction)) -> LuaResult<()> {
    println!("Registered feature {name:?}");
    Ok(())
}
```
Now I need to somehow store the callback inside a HashMap and wonder how I would get a mutable reference to a data structure such that it can be shared and used from this rust function. Is there any easy way?

**Comment by khvzak** [comment, marked as answer]
You can store arbitrary data in Lua itself using the [`app_data`](https://docs.rs/mlua/0.8.9/mlua/struct.Lua.html#method.set_app_data) container.

eg:
```rust
    let map = HashMap::<String, RegistryKey>::new();
    lua.set_app_data(map);
```

However for your case probably would be easier to store function inside registry with given name:
```rust
lua.set_named_registry_value(&name, callback)?;

// Call previously saved function
lua.named_registry_value::<Function>(&name)?.call(())?;
```

**Reply by DasLixou** [reply]
Thanks! AppData seems like the thing I've been looking for. What does registry do exactly? Like mark LuaRefs with a name or what?

**Reply by khvzak** [reply]
You can read about Lua registry in https://www.lua.org/manual/5.4/manual.html#4.3
It's a special Lua table to store arbitrary values.

**Reply by DasLixou** [reply]
I run into a problem.. I wanted to store them into AppData in a HashMap but LuaFunction needs a lifetime. so I gave it one. now i have the problem where app_data_mut wants the AppData to be : 'static... any idea how to get around this?
Here the code and error:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=efec9b2b1f17175fafc15c6f512f9391
```rust
error: lifetime may not live long enough
  --> src\main.rs:47:9
   |
44 | fn feature<'l>(lua: &'l Lua, (name, callback): (String, LuaFunction<'l>)) -> LuaResult<()> {
   |            -- lifetime `'l` defined here
...
47 |     lua.app_data_mut::<Features<'l>>()
   |         ^^^^^^^^^^^^ requires that `'l` must outlive `'static`

error: could not compile `bonzaii` (bin "bonzaii") due to previous error
```

**Reply by khvzak** [reply]
You need to create a `'static` reference to your function in registry.
See [`Lua::create_registry_value()`](https://docs.rs/mlua/0.8.9/mlua/struct.Lua.html#method.create_registry_value)
